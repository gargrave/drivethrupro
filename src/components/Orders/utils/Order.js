/*
A simple utility class for keeping track of details relating to an order.

If an existing Order object is passed, its data will be copied into this one;
this is useful for editing existing orders while maintaining the data (particularly the ID).

If nothing is passed in, a new Order will be created with default values.
*/

// A simple "primary key" mock for each object (i.e. to mock the functionality of a DB)
let nextID = 1

export default class Order {
  constructor (otherOrder) {
    if (otherOrder) {
      this.id = otherOrder.id
      this.items = otherOrder.items
      this.totalItems = otherOrder.totalItems
      this.totalPrice = otherOrder.totalPrice
    } else {
      this.id = nextID++
      this.items = []
      this.totalItems = 0
      this.totalPrice = 0
    }
  }

  addItem (orderItem) {
    this.items.push(orderItem)
    this.totalPrice += orderItem.totalPrice
    this.totalItems += orderItem.quantity
  }

  removeItem (orderItem) {
    if (this.items.find(i => i.id === orderItem.id)) {
      this.items = this.items.filter(i => i.id !== orderItem.id)
      this.totalPrice -= orderItem.totalPrice
      this.totalItems -= orderItem.quantity
    }
  }
}
