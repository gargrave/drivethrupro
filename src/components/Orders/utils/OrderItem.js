/*
A simple utility class for keeping track of a single line item within an order.
*/
let nextID = 1

export default class OrderItem {
  constructor (name, quantity, size) {
    this.id = nextID++
    this.name = `${name} (${size.name})`
    this.quantity = quantity
    this.pricePerUnit = size.price
    this.totalPrice = this.quantity * this.pricePerUnit
  }
}
