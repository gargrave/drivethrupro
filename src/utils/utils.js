export function displayAsDollars (cents) {
  if (!Number.isInteger(cents)) {
    throw new TypeError('"cents" must be an Integer.')
  }

  const prefix = cents < 0 ? '-' : ''
  const absValue = Math.abs(cents * 0.01)
  return `${prefix}$${absValue.toFixed(2)}`
}
