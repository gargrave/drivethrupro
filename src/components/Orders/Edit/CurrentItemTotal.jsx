import React from 'react'
import { number, object } from 'prop-types'

import { displayAsDollars } from '../../../utils/utils'

import './OrderItemForm.css'

const OrderForm = props => {
  const ppu = props.item.sizes[props.sizeIdx].price
  const itemTotal = ppu * props.quantity

  return (
    <div className="item-total-display">
      <h4 style={{ display: 'inline' }}>Add an Item</h4>
      <p>
        Item Total:&nbsp;&nbsp;{props.quantity}x @ {displayAsDollars(ppu)} ={' '}
        <strong>{displayAsDollars(itemTotal)}</strong>
      </p>
    </div>
  )
}

OrderForm.propTypes = {
  item: object.isRequired,
  sizeIdx: number.isRequired,
  quantity: number.isRequired
}

export default OrderForm
