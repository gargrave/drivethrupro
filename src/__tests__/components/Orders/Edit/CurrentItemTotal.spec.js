import React from 'react'
import { shallow } from 'enzyme'

import { displayAsDollars } from '../../../../utils/utils'
import CurrentItemTotal from '../../../../components/Orders/Edit/CurrentItemTotal'

import menu from '../../../../data/menu-items'

describe('CurrentItemTotal', () => {
  let props
  let component

  beforeEach(() => {
    props = {
      item: menu[0],
      sizeIdx: 0,
      quantity: 2
    }

    component = shallow(<CurrentItemTotal {...props} />)
  })

  test('renders correctly', () => {
    expect(component).toMatchSnapshot()
  })

  test('displays the correct quantity', () => {
    const text = component.find('.item-total-display p').text()
    expect(text.indexOf(`${props.quantity}x`)).not.toEqual(-1)
  })

  test('displays the correct price per unit', () => {
    const text = component.find('.item-total-display p').text()
    const expected = displayAsDollars(props.item.sizes[props.sizeIdx].price)
    expect(text.indexOf(`${expected}`)).not.toEqual(-1)
  })

  test('displays the correct total price (ppu * quantity)', () => {
    const text = component.find('.item-total-display p').text()
    const total = props.item.sizes[props.sizeIdx].price * props.quantity
    const expected = displayAsDollars(total)
    expect(text.indexOf(`${expected}`)).not.toEqual(-1)
  })
})
