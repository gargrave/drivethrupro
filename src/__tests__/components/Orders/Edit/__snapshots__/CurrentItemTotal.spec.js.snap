// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentItemTotal renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="item-total-display"
>
    <h4
        style={
            Object {
                "display": "inline",
              }
        }
    >
        Add an Item
    </h4>
    <p>
        Item Total:  
        2
        x @ 
        $0.99
         =
         
        <strong>
            $1.98
        </strong>
    </p>
</div>,
  "nodes": Array [
    <div
      className="item-total-display"
>
      <h4
            style={
                  Object {
                        "display": "inline",
                      }
            }
      >
            Add an Item
      </h4>
      <p>
            Item Total:  
            2
            x @ 
            $0.99
             =
             
            <strong>
                  $1.98
            </strong>
      </p>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <OrderForm
        item={
                Object {
                        "id": 0,
                        "name": "Hamburger",
                        "sizes": Array [
                          Object {
                            "name": "Regular",
                            "price": 99,
                          },
                        ],
                      }
        }
        quantity={2}
        sizeIdx={0}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "item": Object {
            "id": 0,
            "name": "Hamburger",
            "sizes": Array [
              Object {
                "name": "Regular",
                "price": 99,
              },
            ],
          },
          "quantity": 2,
          "sizeIdx": 0,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="item-total-display"
>
          <h4
                    style={
                              Object {
                                        "display": "inline",
                                      }
                    }
          >
                    Add an Item
          </h4>
          <p>
                    Item Total:  
                    2
                    x @ 
                    $0.99
                     =
                     
                    <strong>
                              $1.98
                    </strong>
          </p>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="item-total-display"
>
          <h4
                    style={
                              Object {
                                        "display": "inline",
                                      }
                    }
          >
                    Add an Item
          </h4>
          <p>
                    Item Total:  
                    2
                    x @ 
                    $0.99
                     =
                     
                    <strong>
                              $1.98
                    </strong>
          </p>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <OrderForm
    item={
        Object {
            "id": 0,
            "name": "Hamburger",
            "sizes": Array [
              Object {
                "name": "Regular",
                "price": 99,
              },
            ],
          }
    }
    quantity={2}
    sizeIdx={0}
/>,
}
`;
